### ISSUE TITLE
String in sub expression incorrectly parsed

### ISSUE THREAD

**Issue by colinardo** [PowerShell/PowerShell#4543]
Hello everyone,
got a possible interpreter bug explained in detail first in the following forum
https://social.technet.microsoft.com/Forums/windowsserver/en-US/aeff1790-7e9a-4dfa-b35c-212a7f4b78b0/subexpression-with-additional-parentheses-string-inside-a-string-with-quotation-marks?forum=winserverpowershell

Steps to reproduce
------------------
Enter the following code in the console or in a script:

`"$(')')"` or `"$('(')"` or `"$(")")"` and so on.

Expected behavior
-----------------

Should output a single closing parenthesis **)** in the console

`)`

Actual behavior
---------------

Complains about missing parenthesis, although the parenthesis is enclosed in single quotation marks.
If you add an additional parenthesis so that the parethesis are balanced it's OK but this is not expected.

`"$('()')"`

Environment data
----------------

Could be reproduced in all Powershell-Versions up to the latest available.

Regards
colinardo

**Comment by mklement0** [comment]
Good find; it seems that any  unbalanced `(` or `)` instances inside a _quoted_string_ inside `$(...)` cause the problem.

The following commands all break unexpectedly:

```powershell
"$(')')"  # your example
"$('(')"  # opening parenthesis too
"$("(")"  # ditto with double quotes
```

Conversely, the following does _not_ break, even though it should:

```powershell
"$("`")"  # !! Unexpectedly doesn't break, even though "`" is not a well-formed string.
```

**Comment by briantist** [comment]
This looks related to #3039, and to [this question on SO](https://stackoverflow.com/q/47877162/3905079).

**Comment by lzybkr** [comment]
This is related to #3039 in a way, but I think this issue could be fixed whereas I think #3039 probably can't.

First, there is a workaround - put an opening paren in a comment like this:

```powershell
"$(<#(#>')')"
```

Yes it's ugly. There are better alternatives, like using a format string, but I had to mention this for completeness.

A fix for this issue might be as simple as keeping a stack of expected delimiters, and only counting the opening/closing paren when the top of this stack is a paren.

The delimiters pairs that must be handled include `<# #>`, `' '`, `" "`, `@" "@`, `@' '@` and of course `( )`. I think it's safe to ignore `[ ]` and `{ }` here.

It is important to ignore escape characters though, e.g. the following examples are/should be valid with and without matched delimiters:

```powershell
echo a`'`(b
echo a`'`(`'b
"$(echo a`'`(b)'
"$(echo a`'`(`'b)"
```

**Comment by briantist** [comment]
Really great info provided here, thanks @lzybkr .

**Comment by msftrncs** [comment]
See also 
https://github.com/PowerShell/vscode-powershell/issues/1401

A comment nested inside a subexpression can expose issues with parenthesis as well.

**Comment by msftrncs** [comment]
I ran in to this, unmatched parenthesis because of a regular expression, was able to do this:

from https://gist.github.com/msftrncs/bff8c6c5e28ff92a19efb8a5556a4238, both lines are enclosed in the same string subexpression, and line 37 also possesses its own string subexpression as well.

Line 25
```PowerShell
                if ($_ -match '^(?:[@#<>]|[1-6]>)|[\s`|&;,''"\u2018-\u201E{}()]|\$[{(\w:$^?]') { #)
```

Line 37
```PowerShell

                "$QuotedWith$($_ -replace '["\u201C-\u201E`]', '$0$0' -replace '\$(?=[{(\w:$^?])'<#)#>, '`$0')$QuotedWith"
```

**Comment by UberKluger** [comment]
This problem ***is*** very similar to #3039, #10045 and #10413. Again, the problem is that the parser is examining the contents of the sub-expression instead of treating it as a separate syntactic context. The following is a theory about what is happening here that seems consistent with the examples given and other tests performed. It could be wrong in its detail but does explain the behaviour.

When the parser encounters a subexpression in an expandable string, it determines the extent of that subexpression by counting balanced parentheses. It then builds a substring of what it thinks is the subexpression (the bit from the start "$(" to what it thinks is the end ")", inclusive) and notes where the remainder of the expandable string resumes. The substring is then parsed (as a naked subexpression) to evaluate the expression. Provided that it is possible to parse a valid subexpression beginning at the first character of the substring, anything after that expression is ignored. The subexpression parser assumes that the string parser knows where the subexpression ends and the string resumes (as I said, the theory *explains* the behaviour but might not be exactly what is happening). Once the sub-expression is parsed, the parsing of the expandable string resumes at what the string parser *thinks* is the character following the sub-expression.

Explanation of included examples (using PS ISE 5.1.18362.145 to get errors instead of continuation prompt for "incomplete" input):
```
"$(')')"
The string is missing the terminator: '.
 # parser thinks subexpression is $(') which is invalid.

"$('(')"
$(subexpression) is missing the closing ')'.
# quoted open parenthesis causes parser to expect a second close parenthesis as the subexpression terminator
# which it doesn't find.

"$(")")"
The string is missing the terminator: ".
 # parser thinks subexpression is $(") which is invalid.

"$("`")"  # !! Unexpectedly doesn't break, even though "`" is not a well-formed string.
# This is actually the same as #10045. The `" is treated as an escaped " and replaced before the subexpression is
# parsed. Thus the string becomes
"$("")"
# Now a well formed subexpression. Note: #10413 does not apply, only the `" is replaced. The resulting "" is not
# subject to further replacement as evidenced by the fact that it works ( $(") wouldn't ) and also the following
# example
"$("`"`"`")"
# is changed to
"$("""")"
# producing a single ", i.e. the subexpression becomes an expandable string containing a single escaped "
```
Further examples consistent with the theory,
```
PS > "$("foo("))bar"
foo(bar
PS > "$("foo((")))bar"
foo((bar
PS > "$("foo(")xxxxxxxxxxxx)xbar"
foo(xbar
```
Note how the subexpression correctly evaluates but everything between the *actual* end of the subexpression and the character after what the string parser *thinks* is the *matching* end parenthesis is ignored by both the subexpression (extraneous, assumed to be part of the string) and the string parser (assumed to be part of the subexpression).

 The solution to this problem is simple (to describe). While parsing an expandable string, if a subexpression start sequence is found ( i.e. $( ), begin parsing this as a naked subexpression, totally independent of anything that the string parser is doing. This parsing would succeed if it reached a matching close parenthesis without error. If it succeeds, return a pointer to the next string character after the subexpression. The string parser interpolates the result of the subexpression and then resumes at the character indicated by the subexpression parser.

This will also fix #3039, #10045, #10413 though it might constitute a breaking change (some discussion in #3039). A kludgey way to fix this issue but not the others would be to parse the subexpression in its own context ***after*** replacing all escaped double quotes in the entire string (subexpression included). IMHO: if you're going to fix something, make it work properly by fixing everything that's broken.
